[WT-275]: feat/add options to current breadcrumb

rafijv opened the pull request:
New feature to add a menu of actions to the breadcrumb of the selected folder, the actions added and ready to test are:

- Create new folder
- Share folder
- Get link
- Rename folder
- Move folder
- Download folder
- Move to trash

The expected behavior is that when you click on the last breadcrumb, a dropdown with the previous actions will be displayed and you can perform each of them with the same result that already exists in drive-web.

rafijv committed:
[WT-275] add options to current breadcrumb

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+            >
+              <Menu.Items
+                className={
+                  'absolute mt-6 w-56 rounded-md border border-black border-opacity-8 bg-white py-1.5 drop-shadow'
```
All items texts should be "text-base" (16px)

rafijv replied:
done

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+      {!props.item.active && !props.item.dialog ? (
+        <div className="relative flex items-center">
+          <Menu>
+            {isEditingName ? (
```
There is no inline input to edit folder's name in the design, use the ["Rename dialog"](https://www.figma.com/file/62hpKnJYlxUN77HUECXAlF/Drive-Web?node-id=4649%3A34501&t=PJrB4U0EO0ShnTVU-11) from Figma

rafijv replied:
In the application there is no Rename dialog, it is done with an input when a folder or file has to be renamed. I will create it for this case

aboredvaro replied:
You can re-use ["New folder"](https://github.com/internxt/drive-web/blob/master/src/app/drive/components/CreateFolderDialog/CreateFolderDialog.tsx)'s code, since it follows the standards of usability and accessibility.

rafijv replied:
done

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+    <>
+      {!props.item.active && !props.item.dialog ? (
+        <div className="relative flex items-center">
+          <Menu>
```
Please make sure the menu and its items are fully accessible by keyboard. Check the desired behavior on [HeadlessUI Menu's documentation](https://headlessui.com/react/menu). In the video below you can see the accessibility issue (unable to use keyboard arrows to navigate between the items).

Also folder name should remain in focused style state when the dropdown is open.


https://user-images.githubusercontent.com/34165099/213195577-770d9d46-4d48-46e3-82da-3e8698685511.mov

rafijv replied:
done

sg-gs replied:
Please re-check this @aboredvaro as we want to merge before this week finishes

aboredvaro replied:
Now they are accessible, works as expected

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
-      } cursor-pointer flex-row items-center truncate p-1 font-medium ${isDraggingOverClassNames} 
+    <>
+      {!props.item.active && !props.item.dialog ? (
+        <div className="relative flex items-center">
```
A very sensitive part of the breadcrumb is broken now, be very careful when changing width and text truncate attributes, this causes overflow and breadcrumb item shriking. If you need help with Tailwind ask me.

Check the current behavior of breadcrumb items on [drive.internxt.com](https://drive.internxt.com/app)

<img width="1483" alt="Breadcrumb overflow and item shriking" src="https://user-images.githubusercontent.com/34165099/213184762-45619967-382c-41a1-b215-6cc6c7c3268d.png">

rafijv replied:
done

sg-gs replied:
Please re-check this @aboredvaro as we want to merge before this week finishes

aboredvaro replied:
I've added some corrections with the last commit, now it is working as expected.

aboredvaro commented on src/app/drive/components/MoveItemsDialog/MoveItemsDialog.tsx:
```diff
@@ -154,57 +127,128 @@ const MoveItemsDialog = (props: MoveItemsDialogProps): JSX.Element => {
-      onClose={onClose} closeClass={'hidden'}>
+    <BaseDialog
+      isOpen={isOpen}
+      titleClasses="flex mx-5 text-left font-medium"
```
Title should truncate the item's name, use Tailwind's [truncate](https://tailwindcss.com/docs/text-overflow#truncate). On the other hand good use of the attribute "title" to show the full name of the item in case of overflow.

<img width="537" alt="Title truncate" src="https://user-images.githubusercontent.com/34165099/213185521-5ad44fc5-6794-4ee4-aca3-2c72957e9c79.png">

rafijv replied:
done

aboredvaro replied:
Solved it with tailwind's truncate in the last commit.

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+              >
+                <div className="flex items-center">
+                  {dirtyName}
+                  <CaretDown weight="bold" className="ml-1 h-3 w-3" color="#3A3A3B" />
```
Icon's weight should be `fill`, not `bold`.
Use `text-gray-100` (focus) and `text-gray-80` (enabled) inside className instead of hardcoding the color.

rafijv replied:
done

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+              />
+            ) : (
+              <Menu.Button
+                className={'flex cursor-pointer items-center rounded-md p-1 hover:bg-gray-5 active:bg-gray-5'}
```
Add `p-1 px-1.5` to give 6px on both sides instead of 4px.

rafijv replied:
done

aboredvaro commented on src/app/drive/components/MoveItemsDialog/MoveItemsDialog.tsx:
```diff
@@ -154,57 +127,128 @@ const MoveItemsDialog = (props: MoveItemsDialogProps): JSX.Element => {
-        <div className='ml-5'>{isLoading ? <Spinner className="h-5 w-5 mb-3" /> : <Breadcrumbs items={breadcrumbItems(currentNamePaths)} />}
+      <div className="justify-left w-fill block items-center bg-white py-6 text-left">
+        <div className="ml-5">
+          {isLoading ? <Spinner className="mb-3 h-5 w-5" /> : <Breadcrumbs items={breadcrumbItems(currentNamePaths)} />}
```
Move dialog's breadcrumb logic is attached to drive's breadcrumb logic, the navigation inside the move dialog should not affect drive's navigation current state.

https://user-images.githubusercontent.com/34165099/213187988-566eef89-e8b3-4a7c-8ca6-c4433ef7e096.mov

rafijv replied:
Still working, I can't separate the logic

rafijv replied:
done

aboredvaro replied:
There are still some issues with the move dialog breadcrumb logic:
- When it opens, resets the current breadcrumb state of the user, taking them back to the root folder.
- It also refreshes the main breadcrumb when user goes back to the root folder on the move dialog.
- When **moving** an item, notification should say "Item moved". (But still say "Item restored" when it is used to move an item out of the thrash)

https://user-images.githubusercontent.com/34165099/214880301-3f4f8fd0-d1b6-4a88-bf4c-bc7b1a38ce61.mov

rafijv replied:
done

aboredvaro commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -84,31 +107,222 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+                  </div>
+                </Menu.Item>
+                <div className="my-0.5 mx-3 border-t border-gray-10" />
+                {!isBreadcrumbItemShared ? (
```
Sharing options (🔴 "Share" and "Get link") should not exclude each other, they are different options.

**Options:**
- "Share" → Always (**🔴 [THIS FEATURE](https://www.figma.com/file/62hpKnJYlxUN77HUECXAlF/Drive-Web?node-id=4290%3A31537&t=PJrB4U0EO0ShnTVU-11) IS NOT YET IMPLEMENTED:** It's another feature different to shared links, don't add it yet).
- "Get link" → Only if folder has not a shared link yet.
- "Copy link", "Link settings" and "Delete link" → When the folder has a shared link.

If you have any question about this, please ask Álvaro.

rafijv replied:
done

sg-gs replied:
Please re-check this @aboredvaro as we want to merge before this week finishes

aboredvaro replied:
Working as expected now 👌🏻

aboredvaro reviewed (changes requested):
Please check the changes and if you have any doubt or need help with something ask me 👌🏻

rafijv committed:
[WT-275] fix options to current breadcrumb

rafijv committed:
[WT-275] fix move items dialog breadcrumb

rafijv committed:
Merge branch 'master' into feat/WT-275-breadcrumb-current-folder-options-dropdown

CandelR commented on src/app/drive/components/EditFolderNameDialog/EditFolderNameDialog.tsx:
```diff
@@ -0,0 +1,121 @@
+export default connect((state: RootState) => ({
+  user: state.user.user,
+  neededFolderId: storageSelectors.currentFolderId(state),
+}))(CreateFolderDialog);
```
I don't think it is necessary to connect this Dialog to redux. This global status data is not being used in the Dialog.

rafijv replied:
Correct, I'll change it

CandelR commented on src/app/drive/components/MoveItemsDialog/MoveItemsDialog.tsx:
```diff
@@ -158,57 +129,132 @@ const MoveItemsDialog = (props: MoveItemsDialogProps): JSX.Element => {
+    itemsToMove.length > 1
+      ? itemsToMove.length + ' items'
+      : '"' +
+        (itemsToMove[0]?.name.length > 20 ? itemsToMove[0]?.name.slice(0, 20) + '...' : itemsToMove[0]?.name) +
```
With tailwind truncate that has comented aboredvaro it is necessary the slice logic?

rafijv replied:
I tried it but it doesn't work with truncate

aboredvaro replied:
It can work with truncate, it was using an old component `<BaseDialog />` which is not very flexible. I just switched `<BaseDialog />` with `<Modal />` and refactored the UI of the dialog.

CandelR commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -121,31 +145,249 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
+      if (selectedFolder) foldersList.push(selectedFolder as DriveItemData);
+    }
+    return foldersList;
+  };
```
I think that not exists the possibility to finde more than one item with same id and name. 
If so, an alternative to the findCurrentFolder implementation that saves traversing the entire list and returns a single item, not an array.
```suggestion
  const findCurrentFolder = (currentBreadcrumb) => allItems.find((item) => item.id === currentBreadcrumb.id && item.name === currentBreadcrumb.name);
```

rafijv replied:
I do it this way because of the type of object (allItems) that contains arrays, only with the 'find' it doesn't work

CandelR replied:
ah okay! 
In that case if you are only going to return one item at most I think it would be better to return the value and not an array with the item in it. 
Or it is any reason to store it in an array?

rafijv replied:
Yes, it is the type of data expected in moveItemsToTrash, downloadItemsThunk, setItemsToMove

CandelR replied:
Okay, but that's no reason to choose that data type for the function to return. 
A function is usually better to return the type according to the expected behaviour. In this case, if we expect it to return a single item (the current folder), it should return the data type that corresponds to that single item, and not a list with one item. 
To send it as an argument through the functions you mention you can add the list when you are going to pass it `[item]`. 
I'm sorry if I insist a lot on something that doesn't seem important, but it's to keep the code as consistent and clean as possible.  🙏

rafijv replied:
Maybe I have not explained well or it is not clear but the function is created to reuse the code in these cases.
If I change the function to return a result that then I have to manipulate in several places in the same way I am repeating code to generate the expected result, I think it is better to do it only once inside the function and get the expected result.
But if you continue thinking that it is better that the result of the function is not an array I change it. 😉

CandelR commented on test/e2e/cypress/utils/photos.mock.ts:
```diff
@@ -3,157 +3,189 @@ const URL_BRIDGE = Cypress.env('REACT_APP_STORJ_BRIDGE');
-    PHOTO2 = 'PHOTO2',
-    MULTIPLE = 'MULTIPLE'
-};
+  EMPTY = 'EMPTY',
```
Is there any change in the photos tests and mocks? I only see the formatting of the code. 
Just to make sure I don't miss anything.

rafijv replied:
It's just code formatting

rafijv committed:
[WT-275] fix EditFolderNameDialog export

aboredvaro commented on src/app/drive/components/EditFolderNameDialog/EditFolderNameDialog.tsx:
```diff
@@ -0,0 +1,116 @@
+
+  useEffect(() => {
+    if (isOpen) {
+      setFolderName(currentBreadcrumb[0].name);
```
This is causing the name not to be updated correctly when reopening the rename dialog after renaming the folder.
- Throws a generic error when trying to rename it again. Just rename it to the input value (even if it is the same name, it is not an error).
- Update the input default value properly, if you want to still use the breadcrumb to get the current folder name make sure it is updated.
- When the input shows an error it also shows a notification, don't show notifications with errors on this dialog.

https://user-images.githubusercontent.com/34165099/214904663-68e0d7e5-7e9c-4f50-b584-2d9b39d5e6f9.mov

rafijv replied:
done

aboredvaro commented on src/app/shared/components/Breadcrumbs/Breadcrumbs.tsx:
```diff
@@ -50,6 +51,7 @@ export default function Breadcrumbs(props: BreadcrumbsProps): JSX.Element {
               item={items[i]}
               isHiddenInList
               totalBreadcrumbsLength={items.length}
+              items={items}
```
Passing `JSX.Element[]` as list items is causing some click/hover problems (and is not a good practice).
On the video you can see that if the user don't click exactly on the text it looks like its not working.

https://user-images.githubusercontent.com/34165099/214921232-9bd1b252-385e-4a95-adca-d43719a88598.mov

rafijv replied:
done

aboredvaro committed:
Fixed some UI/UX and accessibility issues

- Fixed breadcrumb truncate issues
- Fixed move dialog accessibility and UX
- Fixed breadcrumb item menu issues

rafijv committed:
[275] fix MoveItemsDialog

rafijv committed:
[275] fix rename dialog

CandelR commented on src/app/shared/components/Breadcrumbs/BreadcrumbsItem/BreadcrumbsItem.tsx:
```diff
@@ -366,7 +364,7 @@ const BreadcrumbsItem = (props: BreadcrumbsItemProps): JSX.Element => {
           className={`max-w-fit flex ${
             props.item.isFirstPath ? 'flex-shrink-0' : 'min-w-breadcrumb flex-1'
-          } cursor-pointer flex-row items-center truncate p-1 font-medium ${isDraggingOverClassNames}
+          } cursor-pointer flex-row items-center truncate py-1.5 px-3 font-medium ${isDraggingOverClassNames}
```
Be careful with style changes, make sure you don't break other elements.

Before change:
![before](https://user-images.githubusercontent.com/57987155/215454937-b98a6994-686d-4080-9f56-ce095cd1173d.png)
After change:
![after](https://user-images.githubusercontent.com/57987155/215454940-ebac2194-f1c0-4863-836c-c432c8f35567.png)

rafijv replied:
done

rafijv committed:
[275] fix breadcrumb dropdown
